File: server/seed.py
from sqlalchemy.orm import Session
from database import SessionLocal, engine, Base
from models import (
    Organisation, Client, Policy, Agreement, Instalment,
    OrganisationStatusEnum, AgreementStatusEnum, InstalmentStatusEnum
)
from datetime import datetime, timedelta
from decimal import Decimal

def seed_database():
    print("🌱 Seeding database...")
    
    # Create tables
    Base.metadata.create_all(bind=engine)
    
    db = SessionLocal()
    
    try:
        # Create organisation
        org = Organisation(
            name="Demo Insurance Brokers Ltd",
            status=OrganisationStatusEnum.ACTIVE
        )
        db.add(org)
        db.flush()
        print(f"✅ Created organisation: {org.name}")
        
        # Create client
        client = Client(
            organisation_id=org.id,
            first_name="John",
            last_name="Doe",
            email="john.doe@example.com",
            phone="+44 20 7123 4567",
            address_line1="123 High Street",
            city="London",
            postcode="SW1A 1AA"
        )
        db.add(client)
        db.flush()
        print(f"✅ Created client: {client.first_name} {client.last_name}")
        
        # Create policy
        policy = Policy(
            organisation_id=org.id,
            client_id=client.id,
            insurer="Aviva",  # Changed from insurer_name
            product_type="Motor",
            policy_number="POL-2024-001",
            start_date=datetime(2024, 1, 1),  # Changed from inception_date
            end_date=datetime(2024, 12, 31),  # Changed from expiry_date
            premium_amount_pennies=120000  # Changed from gross_premium (1200.00 * 100)
        )
        db.add(policy)
        db.flush()
        print(f"✅ Created policy: {policy.policy_number}")
        
        # Create agreement
        principal_amount_pennies = 120000  # £1200 in pennies
        term_months = 12
        apr_bps = 995
        broker_fee_bps = 200  # 2% broker fee
        monthly_rate = (apr_bps / 10000) / 12
        monthly_payment = float(principal_amount_pennies / 100) * monthly_rate / (1 - pow(1 + monthly_rate, -term_months))
        
        agreement = Agreement(
            organisation_id=org.id,
            client_id=client.id,
            policy_id=policy.id,
            principal_amount_pennies=principal_amount_pennies,
            apr_bps=apr_bps,
            term_months=term_months,
            broker_fee_bps=broker_fee_bps,
            status=AgreementStatusEnum.ACTIVE,
            signed_at=datetime(2024, 1, 1),
            activated_at=datetime(2024, 1, 2)
        )
        db.add(agreement)
        db.flush()
        print(f"✅ Created agreement: {agreement.id}")
        
        # Create instalments
        for i in range(term_months):
            due_date = datetime(2024, 1, 1) + timedelta(days=i * 30)
            amount_pennies = int(round(monthly_payment * 100))  # Convert to pennies

            instalment = Instalment(
                agreement_id=agreement.id,
                sequence_number=i + 1,
                due_date=due_date,
                amount_pennies=amount_pennies,
                status=InstalmentStatusEnum.PAID if i < 3 else InstalmentStatusEnum.UPCOMING
            )
            db.add(instalment)

        print(f"✅ Created {term_months} instalments")

        db.commit()
        print("🎉 Seeding completed successfully!")
        
    except Exception as e:
        print(f"❌ Seeding failed: {e}")
        db.rollback()
        raise
    finally:
        db.close()

if __name__ == "__main__":
    seed_database()

File: server/prisma/seed.ts
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();

async function main() {
  console.log('🌱 Seeding database...');

  // Create organisation
  const org = await prisma.organisation.create({
    data: {
      name: 'Demo Insurance Brokers Ltd',
      status: 'ACTIVE',
    },
  });
  console.log('✅ Created organisation:', org.name);

  // Create broker admin
  const brokerAdmin = await prisma.brokerUser.create({
    data: {
      organisationId: org.id,
      role: 'BROKER_ADMIN',
      email: 'admin@demobroker.com',
      name: 'Sarah Johnson',
    },
  });
  console.log('✅ Created broker admin:', brokerAdmin.email);

  // Create broker
  const broker = await prisma.brokerUser.create({
    data: {
      organisationId: org.id,
      role: 'BROKER',
      email: 'broker@demobroker.com',
      name: 'Michael Smith',
    },
  });
  console.log('✅ Created broker:', broker.email);

  // Create client
  const client = await prisma.client.create({
    data: {
      organisationId: org.id,
      firstName: 'John',
      lastName: 'Doe',
      email: 'john.doe@example.com',
      phone: '+44 20 7123 4567',
      dob: new Date('1985-06-15'),
      address: '123 High Street, London, SW1A 1AA',
    },
  });
  console.log('✅ Created client:', `${client.firstName} ${client.lastName}`);

  // Create policy
  const policy = await prisma.policy.create({
    data: {
      organisationId: org.id,
      clientId: client.id,
      insurerName: 'Aviva',
      productType: 'Motor',
      policyNumber: 'POL-2024-001',
      inceptionDate: new Date('2024-01-01'),
      expiryDate: new Date('2024-12-31'),
      grossPremium: 1200.00,
    },
  });
  console.log('✅ Created policy:', policy.policyNumber);

  // Create agreement
  const principalAmount = 1200.00;
  const termMonths = 12;
  const aprBps = 995; // 9.95%
  const monthlyRate = (aprBps / 10000) / 12;
  const monthlyPayment = principalAmount * monthlyRate / (1 - Math.pow(1 + monthlyRate, -termMonths));
  
  const agreement = await prisma.agreement.create({
    data: {
      organisationId: org.id,
      clientId: client.id,
      policyId: policy.id,
      principalAmount,
      aprBps,
      termMonths,
      status: 'ACTIVE',
      startDate: new Date('2024-01-01'),
      endDate: new Date('2024-12-31'),
      outstandingAmount: principalAmount,
    },
  });
  console.log('✅ Created agreement:', agreement.id);

  // Create instalments
  const instalments = [];
  for (let i = 0; i < termMonths; i++) {
    const dueDate = new Date('2024-01-01');
    dueDate.setMonth(dueDate.getMonth() + i);
    
    instalments.push({
      agreementId: agreement.id,
      sequence: i + 1,
      dueDate,
      amountDue: Math.round(monthlyPayment * 100) / 100,
      amountPaid: i < 3 ? Math.round(monthlyPayment * 100) / 100 : 0,
      status: i < 3 ? 'PAID' : 'UPCOMING' as const,
    });
  }

  await prisma.instalment.createMany({
    data: instalments,
  });
  console.log(`✅ Created ${instalments.length} instalments`);

  // Create agreement event
  await prisma.agreementEvent.create({
    data: {
      agreementId: agreement.id,
      type: 'AGREEMENT_CREATED',
      actorType: 'BROKER',
      meta: {
        brokerId: broker.id,
        brokerName: broker.name,
      },
    },
  });

  // Create commission line
  await prisma.commissionLine.create({
    data: {
      agreementId: agreement.id,
      type: 'ORIGINATION',
      amount: 60.00, // 5% of principal
      currency: 'GBP',
    },
  });

  console.log('🎉 Seeding completed successfully!');
}

main()
  .catch((e) => {
    console.error('❌ Seeding failed:', e);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
